libkbfs/key_manager.go

```diff
@@ -733,24 +757,32 @@ func (km *KeyManagerStandard) Rekey(ctx context.Context, md *RootMetadata, promp
 		}
 	}()
 
+	// From this point on, if we return true for mdChanged with a
+	// nil error or RekeyIncompleteError{}, we must call
+	// km.finalizeRekey(md) first.
+
+	defer func() {
+		if mdChanged && (err == nil || err == RekeyIncompleteError{}) {
+			finalizeErr := km.finalizeRekey(
+				ctx, md, tlfCryptKey, promptPaper)
+			if finalizeErr != nil {
+				mdChanged = false
+				cryptKey = nil
+				err = finalizeErr
+			}
+		}
+	}()
+
 	if !isWriter {
 		if len(newReaderUsers) > 0 || addNewWriterDevice || incKeyGen {
 			// If we're a reader but we haven't completed all the work, return
 			// RekeyIncompleteError.
 			return addNewReaderDeviceForSelf, nil, RekeyIncompleteError{}
 		}
 		// Otherwise, there's nothing left to do!
-		if err := md.finalizeRekey(km.config.Crypto(),
-			kbfscrypto.TLFCryptKey{}, tlfCryptKey); err != nil {
```

Comment:
Okay, reworked this by moving all the TLFCryptKey encryption logic into AddKeyGeneration, which is where it should be anyway.